### PR TITLE
Lightning Trainer for MA OSS

### PR DESCRIPTION
**Pull request by dkurra**
PyTorch Lightning trainer + MovieLens demo

## Summary

One-time snapshot of the internal Uber Michelangelo `sdk/trainer/torch/pytorch_lightning/` package into OSS, plus a runnable MovieLens-100k NCF example that exercises it end-to-end. Per the [original Slack thread](https://uber.slack.com/archives/C036Y1N3MND/p1778634921605049): snapshot only, no upstream sync.

Goal: give external users a thin Ray Train wrapper around PyTorch Lightning that the Michelangelo team uses internally, without the closed-source orchestration layer (`workflow/`, `model_manager/`, `artifact/`, CFS filesystem, M3 metrics) it normally sits behind.

Screenshot 2026-05-20 at 4 33 58 PM

## What's in the snapshot

8 files, ~1,400 LOC under `python/michelangelo/lib/trainer/torch/`:

| File | LOC | Role |
|---|---:|---|
| `pytorch_lightning/lightning_trainer.py` | 184 | Public API: `LightningTrainer` (Ray `TorchTrainer` subclass), `LightningTrainerParam`, `CometParam`, `LightningTrainerWithStateDict` |
| `pytorch_lightning/_private/util.py` | 452 | `_train_loop_per_worker` — the per-Ray-worker training loop. Builds model, applies warm-start, resolves strategy/logger/callbacks, wraps Ray dataset shards, fits via `pl.Trainer` |
| `pytorch_lightning/_private/callbacks.py` | 114 | `RayTrainReportCallback` + per-node variant — bridges Lightning's epoch-end hooks to `ray.train.report(...)` for checkpointing |
| `pytorch_lightning/schema.py` | 76 | Warm-start types: `TransferLearningSpec`, `IncrementalTrainingSpec`, `ModelSpec`, `TrainingType`/`LearningMode` enums |
| `data_collate_functions.py` | 259 | Ray Data → torch.Tensor collation. Default `literal_eval_data_collate_function` + ragged-list padding |
| `_numpy_utils.py` | 167 | Inlined `pad_ragged_tensor` + sentinel-dtype helpers (replaces internal `uber.shared.utils.numpy_utils.*`) |
| `utils.py` | 131 | Standalone memory-footprint estimators for transformers + generic `nn.Module` (used by callers to size batches) |

### Execution flow (one Ray worker)

```
LightningTrainer(...).train()
  └── ray.train.torch.TorchTrainer spawns N workers
      └── _train_loop_per_worker(config)        ◄── _private/util.py
          ├── model = create_model_fn(**kwargs)
          ├── _load_weights_from_path / _apply_layer_freeze   (if specs set)
          ├── strategy  = _resolve_strategy()    # RayDDPStrategy by default
          ├── logger    = _resolve_logger(comet_param OR lightning_trainer_kwargs["logger"])
          ├── callbacks = [..., RayTrainReportCallback]   ◄── _private/callbacks.py
          ├── data_iter = wrap(get_dataset_shard("train"), collate=literal_eval_data_collate_function)
          ├── pl_trainer = ray.train.lightning.prepare_trainer(pl.Trainer(...))
          └── pl_trainer.fit(model, train_iter, val_iter)
              └── each epoch end → ray.train.report(metrics, checkpoint)
```

## What's intentionally stripped from the internal version

To make the snapshot standalone:

- **ArtifactStore epoch-resume** — internal trainer polls CFS for the latest checkpoint and resumes mid-run. OSS users start from local paths only.
- **M3 metric counters** — `MichelangeloStatsLogger`, `CANVAS_V2_TRAINING_JOB_RESULT_METRIC`, `CHECKPOINT_UPLOAD_METRIC` all removed.
- **`register_cfs_into_fsspec`** — Uber-internal filesystem; OSS users use local / S3 / GCS via standard fsspec.
- **Rich `UserException`** — replaced with a minimal local class.
- **Internal `numpy_utils` / `sentinel` packages** — relevant helpers inlined into `_numpy_utils.py`.

Result: zero `uber.*` imports in the snapshot.

## MovieLens-100k demo (`python/examples/movielens/`)

Smallest viable smoke test for the trainer. Trains a tiny NCF (~92K params) on CPU with one Ray Train worker, 3 epochs, ~5s wall time on the validation run.

| File | Role |
|---|---|
| `data.py` | Downloads ml-100k (~5 MB), remaps user/item IDs to dense indices, returns Ray Datasets. Has a GitHub mirror fallback for sandboxed environments. |
| `model.py` | `NCFLightningModule`: user + item embeddings → 2-layer MLP → sigmoid, MSE on `[0,1]`-normalized ratings. |
| `train.py` | Wires up `LightningTrainerParam`, `LightningTrainer`, `RunConfig`, `ScalingConfig`. Optionally enables Comet *or* MLflow tracking via env vars (Comet wins if both set). |
| `README.md` | One-shot run instructions + env-var recipes for each tracking backend. |

What the demo exercises end-to-end:

- Loading `LightningTrainer` + `LightningTrainerParam` from the snapshot.
- `_train_loop_per_worker` for a non-trivial Lightning fit, including `RayTrainReportCallback` epoch checkpointing.
- Default Ray Data → torch tensor collation (no custom `data_collate_fn`).
- Resolving the default `RayDDPStrategy` even with a single worker.
- The `_resolve_logger` Comet path (when Comet env vars are set).
- The `_resolve_logger` pre-built-`pl.Logger` path (when MLflow env vars are set).

## pyproject changes

Added two optional dependencies + a `trainer` extras group:

```toml
comet_ml  = { version = "^3.49.0", optional = true }
deepspeed = { version = "^0.14.0", optional = true }

trainer = ["ray", "torch", "pytorch_lightning", "transformers", "numpy", "comet_ml", "deepspeed"]
```

## How to run

```bash
cd python/
poetry install --extras "trainer example"
python -m examples.movielens.train
```

First run downloads MovieLens-100k to `/tmp/movielens_data/`; checkpoints land in `/tmp/movielens_runs/ncf_movielens100k/`.

Opt-in tracking:

```bash
# Comet
export COMET_API_KEY=... COMET_WORKSPACE=...
python -m examples.movielens.train

# OR MLflow (Comet wins if both env-sets are present)
export MLFLOW_TRACKING_URI=file:///tmp/mlflow_movielens
python -m examples.movielens.train
mlflow ui --backend-store-uri file:///tmp/mlflow_movielens
```

## Test plan

- [x] `python -m examples.movielens.train` completes in ~5s, 3 epochs, train loss 0.063 → 0.054, val checkpoints written.
- [x] MLflow path verified: experiment created, run FINISHED, all metrics + hyperparameters + tags logged.
- [ ] Comet path verified by external reviewer with a Comet workspace.
- [ ] `poetry install --extras "trainer example"` resolves cleanly on a fresh checkout.

## Commits

1. `Snapshot internal Lightning trainer into lib/trainer/torch` — the core port + pyproject deps.
2. `Add MovieLens-100k NCF example exercising lib/trainer/torch` — `examples/movielens/{data,model,train,README}.py`.
3. `examples/movielens: wire optional Comet logging via env vars`.
4. `examples/movielens: add MLflow as an alternate optional tracker`.

## Out of scope (deliberately)

- Sync mechanism with internal repo. This is a one-time snapshot.
- The non-Lightning subpackages of the internal trainer (`comet/`, `custom/`, `huggingface/`).
- CFS / ArtifactStore equivalents — out of scope for the Lightning trainer alone.

### CODE REVIEW

File: python/pyproject.toml [hunk 1]
```diff
@@ -50,6 +50,8 @@ torch = { version = "2.6.0", optional = true }
 pytorch_lightning = { version = "2.2.0", optional = true }
 einops = { version = "0.8.0", optional = true }
 transformers = { version = "4.48.2", optional = true }
+comet_ml = { version = "^3.49.0", optional = true }
```
Comment (badcount) [inline python/pyproject.toml:53]: we don't have comet, can we not use it here?

Comment (dkurra) [reply python/pyproject.toml:53]: Uber uses comet, how would you recommend to have backward compatibility?

**Review by badcount — APPROVED** [review]

File: python/pyproject.toml (hunk 1 above)
Comment (badcount) [reply python/pyproject.toml:53]: good questions,  i think we need to refactor the trainer for different observability. Stamp for now to refactor later
